content/Patterns for maintainable Rails applications. Part I - Ignacio Pérez.md

```diff
@@ -0,0 +1,111 @@
+
+#### Form Objects
+
+A Form Object takes care of the creation of multiple models, attributes mapping and contextual validations. We can define one form object for each form in the frontend. They also give more flexibility to the frontend as we can map fields in the forms to the attributes of the records in the database. Integrity validations should stay on the model, although validations that are only important in the context of the form should go in the form object. In addition, the pattern helps to respect the Single Responsibility Principle by removing logic from our models or controllers.
+
+Any changes needed to the forms in the frontend will be much easier to implement as we should only need to modify the related form object, and any new form related to a model may simply require extending our code.
+
+Continuing the example we used earlier, now the User model has a user_request which Is responsible for knowing the filters the user saves when browsing the store.
+
+https://gist.github.com/PerezIgnacio/e1babd9cab20f79131941c142b89b402
+
+Let’s assume that when the user signs up in our site, we also send a nested user_request with the filters. We receive this parameters in our controller and create both associated records in the database. In this case, the sign up controller may look something like this:
+
+https://gist.github.com/PerezIgnacio/b5e5212852fcbcff5b5e9c8e9bc74a2c
+
+This looks clean, user creation is simple because we have `accept_nested_attributes_for` in its model. However we are coupling our frontend to our database, while also forcing them to send the `user_request` param with `_attributes`. Also, if the user model and its associations gets more complex we may need to include additional code in the controller and, as we know, controllers should be as thin as possible. As an alternative, we can create a Form object, which could look similar to the following:
+
+https://gist.github.com/PerezIgnacio/da03bb1174565f63d392af570f78b592
+
+As you may have noticed, we need to add more code because nested attributes did things for us that we now have to do manually. However, this code is encapsulated in its own object, so it still is preferable over having code in controllers or models. Basically, the form object must be initialized with the params we receive in our controller and we will map to our models attributes. We define then a save method which is in charge of triggering the model persistence logic.
+
+We also include some validations from ActiveModel, checking the presence of some attributes in the form. All models can also be saved in a transaction, so it rollbacks in case any of the models created has an error. Additionally, we may add code that we want to execute after everything is created correctly, such as triggering a notification.
```

Comment:
maybe we can express the benefit of having form validations vs model validations. Form validations validates business logic vs model validations that validates data integrity. https://revs.runtime-revolution.com/validating-form-objects-8058fefc7b89

Reply:
Sounds good, I will add more about it.